README.md
[![Build Status](https://img.shields.io/travis/pry/pry.svg)](https://travis-ci.org/pry/pry)
[![Code Climate](https://img.shields.io/codeclimate/github/pry/pry.svg)](https://codeclimate.com/github/pry/pry)
[![Inline docs](http://inch-pages.github.io/github/pry/pry.svg)](http://inch-pages.github.io/github/pry/pry)

<center>
![The Pry Logo](https://dl.dropbox.com/u/26521875/pry%20stuff/logo/pry_logo_350.png)

© John Mair ([banisterfiend](https://twitter.com/banisterfiend)) 2013<br>

**Please** [DONATE](http://www.pledgie.com/campaigns/15899) to the Pry project - Pry was a **huge** amount of work and every donation received is encouraging and supports Pry's continued development!

**Sponsors**

[Tealeaf Academy](www.gotealeaf.com)<br/>
[Atomic Object](http://www.atomicobject.com/)<br/>
[Hashrocket](http://hashrocket.com/)<br/>
[Intridea](http://intridea.com/)<br/>
[Gaslight](http://gaslight.co/home)<br/>

**Other Resources**

[Skip to the website (recommended)](http://pry.github.com) <br />
[Skip to the wiki](https://github.com/pry/pry/wiki)
</center>

Pry is a powerful alternative to the standard IRB shell for Ruby. It is
written from scratch to provide a number of advanced features,
including:

* Source code browsing (including core C source with the pry-doc gem)
* Documentation browsing
* Live help system
* Open methods in editors (`edit Class#method`)
* Syntax highlighting
* Command shell integration (start editors, run git, and rake from within Pry)
* Gist integration
* Navigation around state (`cd`, `ls` and friends)
* Runtime invocation (use Pry as a developer console or debugger)
* Exotic object support (BasicObject instances, IClasses, ...)
* A Powerful and flexible command system
* Ability to view and replay history
* Many convenience commands inspired by IPython, Smalltalk and other advanced REPLs
* A wide-range number of [plugins](https://github.com/pry/pry/wiki/Available-plugins) that provide remote sessions, full debugging functionality, and more.

Pry also aims to be more than an IRB replacement; it is an
attempt to bring REPL driven programming to the Ruby language. It is
currently not as powerful as tools like [SLIME](http://en.wikipedia.org/wiki/SLIME) for lisp, but that is the
general direction Pry is heading.

Pry is also fairly flexible and allows significant user
[customization](https://github.com/pry/pry/wiki/Customization-and-configuration)
is trivial to set it to read from any object that has a `readline` method and write to any object that has a
`puts` method - many other aspects of Pry are also configurable making
it a good choice for implementing custom shells.

Pry comes with an executable so it can be invoked at the command line.
Just enter `pry` to start. A `.pryrc` file in the user's home directory will
be loaded if it exists. Type `pry --help` at the command line for more
information.

Try `gem install pry-doc` for additional documentation on Ruby Core
methods. The additional docs are accessed through the `show-doc` and
`show-method` commands.

* Install the [gem](https://rubygems.org/gems/pry): `gem install pry`
* Browse the comprehensive [documentation at the official Pry wiki](https://github.com/pry/pry/wiki)
* Read the [YARD API documentation](http://rdoc.info/github/pry/pry/master/file/README.markdown)
* See the [source code](http://github.com/pry/pry)

### Commands

Nearly every piece of functionality in a Pry session is implemented as
a command. Commands are not methods and must start at the beginning of a line, with no
whitespace in between. Commands support a flexible syntax and allow
'options' in the same way as shell commands, for example the following
Pry command will show a list of all private instance methods (in
scope) that begin with 'pa'

    pry(YARD::Parser::SourceParser):5> ls -Mp --grep ^pa
    YARD::Parser::SourceParser#methods: parse  parser_class  parser_type  parser_type=  parser_type_for_filename

### Navigating around state

Pry allows us to pop in and out of different scopes (objects) using
the `cd` command. This enables us to explore the run-time view of a
program or library. To view which variables and methods are available
within a particular scope we use the versatile [ls command.](https://gist.github.com/c0fc686ef923c8b87715)

Here we will begin Pry at top-level, then Pry on a class and then on
an instance variable inside that class:

    pry(main)> class Hello
    pry(main)*   @x = 20
    pry(main)* end
    => 20
    pry(main)> cd Hello
    pry(Hello):1> ls -i
    instance variables: @x
    pry(Hello):1> cd @x
    pry(20):2> self + 10
    => 30
    pry(20):2> cd ..
    pry(Hello):1> cd ..
    pry(main)> cd ..

The number after the `:` in the pry prompt indicates the nesting
level. To display more information about nesting, use the `nesting`
command. E.g

    pry("friend"):3> nesting
    Nesting status:
    0. main (Pry top level)
    1. Hello
    2. 100
    3. "friend"
    => nil

We can then jump back to any of the previous nesting levels by using
the `jump-to` command:

    pry("friend"):3> jump-to 1
    => 100
    pry(Hello):1>

### Runtime invocation

Pry can be invoked in the middle of a running program. It opens a Pry
session at the point it's called and makes all program state at that
point available. It can be invoked on any object using the
`my_object.pry` syntax or on the current binding (or any binding)
using `binding.pry`. The Pry session will then begin within the scope
of the object (or binding). When the session ends the program continues with any
modifications you made to it.

This functionality can be used for such things as: debugging,
implementing developer consoles and applying hot patches.

code:

    # test.rb
    require 'pry'

    class A
      def hello() puts "hello world!" end
    end

    a = A.new

    # start a REPL session
    binding.pry

    # program resumes here (after pry session)
    puts "program resumes here."

Pry session:

    pry(main)> a.hello
    hello world!
    => nil
    pry(main)> def a.goodbye
    pry(main)*   puts "goodbye cruel world!"
    pry(main)* end
    => nil
    pry(main)> a.goodbye
    goodbye cruel world!
    => nil
    pry(main)> exit

    program resumes here.

### Command Shell Integration

A line of input that begins with a '.' will be forwarded to the
command shell. This enables us to navigate the file system, spawn
editors, and run git and rake directly from within Pry.

Further, we can use the `shell-mode` command to incorporate the
present working directory into the Pry prompt and bring in (limited at this stage, sorry) file name completion.
We can also interpolate Ruby code directly into the shell by
using the normal `#{}` string interpolation syntax.

In the code below we're going to switch to `shell-mode` and edit the
`.pryrc` file in the home directory. We'll then cat its contents and
reload the file.

    pry(main)> shell-mode
    pry main:/home/john/ruby/projects/pry $ .cd ~
    pry main:/home/john $ .emacsclient .pryrc
    pry main:/home/john $ .cat .pryrc
    def hello_world
      puts "hello world!"
    end
    pry main:/home/john $ load ".pryrc"
    => true
    pry main:/home/john $ hello_world
    hello world!

We can also interpolate Ruby code into the shell. In the
example below we use the shell command `cat` on a random file from the
current directory and count the number of lines in that file with
`wc`:

    pry main:/home/john $ .cat #{Dir['*.*'].sample} | wc -l
    44

### Code Browsing

You can browse method source code with the `show-method` command. Nearly all Ruby methods (and some C methods, with the pry-doc
gem) can have their source viewed. Code that is longer than a page is
sent through a pager (such as less), and all code is properly syntax
highlighted (even C code).

The `show-method` command accepts two syntaxes, the typical ri
`Class#method` syntax and also simply the name of a method that's in
scope. You can optionally pass the `-l` option to show-method to
include line numbers in the output.

In the following example we will enter the `Pry` class, list the
instance methods beginning with 're' and display the source code for the `rep` method:

    pry(main)> cd Pry
    pry(Pry):1> ls -M --grep re
    Pry#methods: re  readline  refresh  rep  repl  repl_epilogue  repl_prologue  retrieve_line
    pry(Pry):1> show-method rep -l

    From: /home/john/ruby/projects/pry/lib/pry/pry_instance.rb @ line 143:
    Number of lines: 6

    143: def rep(target=TOPLEVEL_BINDING)
    144:   target = Pry.binding_for(target)
    145:   result = re(target)
    146:
    147:   show_result(result) if should_print?
    148: end

Note that we can also view C methods (from Ruby Core) using the
`pry-doc` plugin; we also show off the alternate syntax for
`show-method`:

    pry(main)> show-method Array#select

    From: array.c in Ruby Core (C Method):
    Number of lines: 15

    static VALUE
    rb_ary_select(VALUE ary)
    {
        VALUE result;
        long i;

        RETURN_ENUMERATOR(ary, 0, 0);
        result = rb_ary_new2(RARRAY_LEN(ary));
        for (i = 0; i < RARRAY_LEN(ary); i++) {
            if (RTEST(rb_yield(RARRAY_PTR(ary)[i]))) {
                rb_ary_push(result, rb_ary_elt(ary, i));
            }
        }
        return result;
    }

### Documentation Browsing

One use-case for Pry is to explore a program at run-time by `cd`-ing
in and out of objects and viewing and invoking methods. In the course
of exploring it may be useful to read the documentation for a
specific method that you come across. Like `show-method` the `show-doc` command supports
two syntaxes - the normal `ri` syntax as well as accepting the name of
any method that is currently in scope.

The Pry documentation system does not rely on pre-generated `rdoc` or
`ri`, instead it grabs the comments directly above the method on
demand. This results in speedier documentation retrieval and allows
the Pry system to retrieve documentation for methods that would not be
picked up by `rdoc`. Pry also has a basic understanding of both the
rdoc and yard formats and will attempt to syntax highlight the
documentation appropriately.

Nonetheless, the `ri` functionality is very good and
has an advantage over Pry's system in that it allows documentation
lookup for classes as well as methods. Pry therefore has good
integration with  `ri` through the `ri` command. The syntax
for the command is exactly as it would be in command-line -
so it is not necessary to quote strings.

In our example we will enter the `Gem` class and view the
documentation for the `try_activate` method:

    pry(main)> cd Gem
    pry(Gem):1> show-doc try_activate

    From: /Users/john/.rvm/rubies/ruby-1.9.2-p180/lib/ruby/site_ruby/1.9.1/rubygems.rb @ line 201:
    Number of lines: 3

    Try to activate a gem containing path. Returns true if
    activation succeeded or wasn't needed because it was already
    activated. Returns false if it can't find the path in a gem.
    pry(Gem):1>

We can also use `ri` in the normal way:

    pry(main) ri Array#each
    ----------------------------------------------------------- Array#each
         array.each {|item| block }   ->   array
    ------------------------------------------------------------------------
         Calls _block_ once for each element in _self_, passing that element
         as a parameter.

            a = [ "a", "b", "c" ]
            a.each {|x| print x, " -- " }

         produces:

            a -- b -- c --

### Gist integration

If the `gist` gem is installed then method source or documentation can be gisted to github with the
`gist` command.  The `gist` command is capable of gisting [almost any REPL content](https://gist.github.com/cae143e4533416529726), including methods, documentation,
input expressions, command source, and so on. In the example below we will gist the C source
code for the `Symbol#to_proc` method to github:

    pry(main)> gist -m Symbol#to_proc
    Gist created at https://gist.github.com/5332c38afc46d902ce46 and added to clipboard.
    pry(main)>

You can see the actual gist generated here: [https://gist.github.com/5332c38afc46d902ce46](https://gist.github.com/5332c38afc46d902ce46)

### Edit methods

You can use `edit Class#method` or `edit my_method`
(if the method is in scope) to open a method for editing directly in
your favorite editor. Pry has knowledge of a few different editors and
will attempt to open the file at the line the method is defined.

You can set the editor to use by assigning to the `Pry.editor`
accessor. `Pry.editor` will default to `$EDITOR` or failing that will
use `nano` as the backup default. The file that is edited will be
automatically reloaded after exiting the editor - reloading can be
suppressed by passing the `--no-reload` option to `edit`

In the example below we will set our default editor to "emacsclient"
and open the `Pry#repl` method for editing:

    pry(main)> Pry.editor = "emacsclient"
    pry(main)> edit Pry#repl

### Live Help System

Many other commands are available in Pry; to see the full list type
`help` at the prompt. A short description of each command is provided
with basic instructions for use; some commands have a more extensive
help that can be accessed via typing `command_name --help`. A command
will typically say in its description if the `--help` option is
avaiable.

### Use Pry as your Rails Console

The recommended way to use Pry as your Rails console is to add
[the `pry-rails` gem](https://github.com/rweng/pry-rails) to
your Gemfile. This replaces the default console with Pry, in
addition to loading the Rails console helpers and adding some
useful Rails-specific commands.

If you don't want to change your Gemfile, you can still run a Pry
console in your app's environment using Pry's `-r` flag:

    pry -r ./config/environment

Also check out the [wiki](https://github.com/pry/pry/wiki/Setting-up-Rails-or-Heroku-to-use-Pry)
for more information about integrating Pry with Rails.

### Limitations:

* Tab completion is currently a bit broken/limited this will have a
  major overhaul in a future version.

### Syntax Highlighting

Syntax highlighting is on by default in Pry. If you want to change
the colors, check out the [pry-theme](https://github.com/kyrylo/pry-theme)
gem.

You can toggle the syntax highlighting on and off in a session by
using the `toggle-color` command. Alternatively, you can turn it off
permanently by putting the line `Pry.color = false` in your `~/.pryrc`
file.

### Future Directions

Many new features are planned such as:

* Increase modularity (rely more on plugin system)
* Much improved documentation system, better support for YARD
* Better support for code and method reloading and saving code
* Extended and more sophisticated command system, allowing piping
between commands and running commands in background

### Contact

Problems or questions? file an issue at [github](https://github.com/pry/pry/issues)

### Contributors

Pry is primarily the work of [John Mair (banisterfiend)](http://github.com/banister), for full list
of contributors see the
[CONTRIBUTORS](https://github.com/pry/pry/blob/master/CONTRIBUTORS) file.


CHANGELOG
10/05/2013 version 0.9.12.2 minor fix
* reload-code (with no args) should reload 'current file', issue #920

21/04/2013 version 0.9.12.1 HOTFIX
* Fixes #890, workaround for jruby crash related to http://jira.codehaus.org/browse/JRUBY-7114
* note this fix is just applied to the 0.9.12 tag, not master. However, master has
	this fix applied too.

12/02/2013 version 0.9.12
Bug fixes:
* Fix major bug where commands broke due to slop looking at ARGV
	instead of command parameters (#828)
* Fix bug where pager broke in some situations (#845)
* Fix broken rendering of some docs (#795)
* silence warnings during failed tab-completion attempts
* fix broken prompt when prompt is colored (#822 / #823)
* added reload-method alias for reload-code (for backwards compat)
* Reopen Readline.output if it is not a tty (see 1538bc0990)

Features:
* added pry --gem (see 19bfc13aa)
* show-source now works on commands created with create_command
* add -m (method) -c (class) and -f (file) switches to 'whereami' command
* show-source now falls back to superclass if can't find class code
	(displays warning)
* show-source/show-doc now indicate when -a option is available.

20/01/2013 version 0.9.11.4
* fixed pager bug (wouldn't render color codes in some circumstances)
* added Pry.last_internal_error, useful for devs debugging commands

17/01/2013 version 0.9.11.3
* fix Pry.run_command
* improve `ls` output
* add :requires_gem => "jist" to 'gist' command (so deps can be installed
	via install-command)
* improve help for 'edit' command

16/01/2013 version 0.9.11.2
* minor bug fix for gist on windows, rescuing Jist::ClipboardError
rather than letting the scary error spill out to users and potentially
have them think the gist didnt post.

16/01/2013 version 0.9.11.1
* fixed minor bug in 'gist' command where i neglected to remove
  a call to a non-existent method (no_arg) which was called when
  `gist` is invoked with no parameters

16/01/2013 version 0.9.11
dependency changes:
* upgrade the slop gem to version ~> 3.4
* new optional dependency: Bond (you'll need to perform `gem install bond`). It enables autocompletion if you use Readline. Does not support work for Editline (more info: https://github.com/pry/pry/wiki/FAQ#wiki-readline). Big thanks to cldwalker.

new features:
* #738 basic Ruby 2.0 support
* #732 JRuby 1.7.0+ support
* new reload-code command, using it you can reload code for methods,
	classes, commands, objects and so on. e.g reload-code MyCLass,
	reload-code my_method, reload-code my_obj
* added bond tabcompletion, much more powerful than standard tab
	completion. However, it requires a real 'readline' to work,
	so will not work on standard osx setup (since it uses Editline)
* show-source can now extract source for classes, methods, procs,
	pry commands, and arbitrary objects (it shows the source for the class of
	the object). As a result, show-command is now deprecated
* gist/play/save-file now infer object type without requiring flags,
	e.g play MyClass, play my_file.rb, play my_method
* edit command can now edit most things, including: files, methods,
	classes, commands. As a result the 'edit-method' command is now
	deprecatd.
	e.g edit my_file.rb, edit my_method, edit MyClass
* removed "edit-method", now merged with 'edit'
* amend-line and play now properly indent code added to input buffer.
* #674 added support for require switches chaining (`pry -rubygems -r./a.rb`)
* #695 added ability to customize the name displayed in the prompt
* #716 added `--patch` switch for `edit --ex` command
* #736 respect the `$PAGER` environment variable
* #497 added `disable-pry` command
* removed "show-command" command; use "show-source" command instead
* exec switch chaining (`pry -e ':one' -e ':two'`)
* added two new hooks: "before_eval" and "after_eval"
* added tabcompletion for "show-source", "show-doc" commands and "Array#<tab>" cases
* immediately require gems after gem-install
* added `-l` switch for `ls` command (displays local variables)
* added "gem-open" command
* added "fix-indent" command
* added subcommands API
* exposed the test API for plugin writers (d1489a)
* tablifed ls output
* added `--no-line-numbers` switch for "whereami" command
* added `--lines` switch for "play" command

bug fixes:
* #652 find-method uses single escape instead of double
* #657 fixed blank string delimiters
* #622 fixed "unwanted 'binding_impl_method' local now exists in scratch bindings"
* #645 fixed "edit-method -p changes constant lookup"
* #682 fixed ".pryrc" loading twice when invoked from `$HOME` directory
* #675 fixed pry not remembering initial "pwd"
* #717 fixed multiline object coloring
* #719 fixed "show-method" not supporting `String::new` notation
* #754 fixed "whereami" command not showing correct line numbers
* #751 fixed buggy Cucumber AST output
* #787 fixed `while/until do` loops indentation
* #526 fixed buggy `--no-plugins` switch of pry
* #774 ensure all errors go to the error handler
* fixed ".pryrc" loading with wrong `__FILE__`
* fixed bug when pager doesn't work if "less" is not available on system
* fixed "Control-D" press in nested REPL
* many small improvements of unclear error messages and formatting of documentation

14/07/2012 version 0.9.10
dependency changes:
* #561 upgrade the slop gem to version 3
* #590 move to the jist gem from gist.
* upgrade method_source to 0.8

new features:
* #572 add --hist, -o and -k flags to gist command
* #584 support show-source/doc on methods defined in class-eval
* #585 support show-source/doc on gem methods defined in C
* #596 add --disable-plugin and --select-plugin options
* #597 allow "cd -" to switch between bindings
* #612 add Pry.config.should_load_local_rc to turn off ./.pryrc
* allow running a file of pry input with pry <file>
* support colours in "ri" command
* add before_eval hook
* prompt now gets a lot more data when proc arity is 1

bug fixes &c.
* #554 removed the "req" command
* #567 fix rendering bugs when starting pry

lib/pry.rb
# (C) John Mair (banisterfiend) 2013
# MIT License
#

require 'pp'
require 'pry/helpers/base_helpers'
require 'pry/hooks'

class Pry
  # The default hooks - display messages when beginning and ending Pry sessions.
  DEFAULT_HOOKS = Pry::Hooks.new.add_hook(:before_session, :default) do |out, target, _pry_|
    next if _pry_.quiet?
    _pry_.run_command("whereami --quiet")
  end

  # The default print
  DEFAULT_PRINT = proc do |output, value|
    output_with_default_format(output, value, :hashrocket => true)
  end

  def self.output_with_default_format(output, value, options = {})
    stringified = begin
                    value.pretty_inspect
                  rescue RescuableException
                    nil
                  end

    unless String === stringified
      # Read the class name off of the singleton class to provide a default
      # inspect.
      eig = class << value; self; end
      klass = Pry::Method.safe_send(eig, :ancestors).first
      id = value.__id__.to_s(16) rescue 0
      stringified = "#<#{klass}:0x#{id}>"
    end

    nonce = rand(0x100000000).to_s(16) # whatever

    stringified.gsub!(/#</, "%<#{nonce}")
    # Don't recolorize output with color (for cucumber, looksee, etc.) [Issue #751]
    colorized = if stringified =~ /\e\[/
                  stringified
                else
                  Helpers::BaseHelpers.colorize_code(stringified)
                end

    # avoid colour-leak from CodeRay and any of the users' previous output
    colorized = colorized.sub(/(\n*)\z/, "\e[0m\\1") if Pry.color

    result = colorized.gsub(/%<(.*?)#{nonce}/, '#<\1')
    result = "=> #{result}"if options[:hashrocket]
    Helpers::BaseHelpers.stagger_output(result, output)
  end

  # may be convenient when working with enormous objects and
  # pretty_print is too slow
  SIMPLE_PRINT = proc do |output, value|
    begin
      output.puts value.inspect
    rescue RescuableException
      output.puts "unknown"
    end
  end

  # useful when playing with truly enormous objects
  CLIPPED_PRINT = proc do |output, value|
    output.puts Pry.view_clip(value)
  end

  # Will only show the first line of the backtrace
  DEFAULT_EXCEPTION_HANDLER = proc do |output, exception, _|
    if UserError === exception && SyntaxError === exception
      output.puts "SyntaxError: #{exception.message.sub(/.*syntax error, */m, '')}"
    else
      output.puts "#{exception.class}: #{exception.message}"
      output.puts "from #{exception.backtrace.first}"
    end
  end

  DEFAULT_PROMPT_NAME = 'pry'

  # The default prompt; includes the target and nesting level
  DEFAULT_PROMPT = [
                    proc { |target_self, nest_level, pry|
                      "[#{pry.input_array.size}] #{Pry.config.prompt_name}(#{Pry.view_clip(target_self)})#{":#{nest_level}" unless nest_level.zero?}> "
                    },

                    proc { |target_self, nest_level, pry|
                      "[#{pry.input_array.size}] #{Pry.config.prompt_name}(#{Pry.view_clip(target_self)})#{":#{nest_level}" unless nest_level.zero?}* "
                    }
                   ]

  DEFAULT_PROMPT_SAFE_OBJECTS = [String, Numeric, Symbol, nil, true, false]

  # A simple prompt - doesn't display target or nesting level
  SIMPLE_PROMPT = [proc { ">> " }, proc { " | " }]

  SHELL_PROMPT = [
                  proc { |target_self, _, _| "#{Pry.config.prompt_name} #{Pry.view_clip(target_self)}:#{Dir.pwd} $ " },
                  proc { |target_self, _, _| "#{Pry.config.prompt_name} #{Pry.view_clip(target_self)}:#{Dir.pwd} * " }
                 ]

  # A prompt that includes the full object path as well as
  # input/output (_in_ and _out_) information. Good for navigation.
  NAV_PROMPT = [
                proc do |conf|
                  tree = conf.binding_stack.map { |b| Pry.view_clip(b.eval("self")) }.join " / "
                  "[#{conf.expr_number}] (#{Pry.config.prompt_name}) #{tree}: #{conf.nesting_level}> "
                end,
                proc do |conf|
                  tree = conf.binding_stack.map { |b| Pry.view_clip(b.eval("self")) }.join " / "
                  "[#{conf.expr_number}] (#{ Pry.config.prompt_name}) #{tree}: #{conf.nesting_level}* "
                end,
               ]

  # Deal with the ^D key being pressed. Different behaviour in different cases:
  #   1. In an expression behave like `!` command.
  #   2. At top-level session behave like `exit` command.
  #   3. In a nested session behave like `cd ..`.
  DEFAULT_CONTROL_D_HANDLER = proc do |eval_string, _pry_|
    if !eval_string.empty?
      eval_string.replace('') # Clear input buffer.
    elsif _pry_.binding_stack.one?
      _pry_.binding_stack.clear
      throw(:breakout)
    else
      # Otherwise, saves current binding stack as old stack and pops last
      # binding out of binding stack (the old stack still has that binding).
      _pry_.command_state["cd"] ||= OpenStruct.new # FIXME
      _pry_.command_state['cd'].old_stack = _pry_.binding_stack.dup
      _pry_.binding_stack.pop
    end
  end

  DEFAULT_SYSTEM = proc do |output, cmd, _|
    if !system(cmd)
      output.puts "Error: there was a problem executing system command: #{cmd}"
    end
  end

  # Store the current working directory. This allows show-source etc. to work if
  # your process has changed directory since boot. [Issue #675]
  INITIAL_PWD = Dir.pwd

  # As a REPL, we often want to catch any unexpected exceptions that may have
  # been raised; however we don't want to go overboard and prevent the user
  # from exiting Pry when they want to.
  module RescuableException
    def self.===(exception)
      case exception
        # Catch when the user hits ^C (Interrupt < SignalException), and assume
        # that they just wanted to stop the in-progress command (just like bash etc.)
      when Interrupt
        true
        # Don't catch signals (particularly not SIGTERM) as these are unlikely to be
        # intended for pry itself. We should also make sure that Kernel#exit works.
      when *Pry.config.exception_whitelist
        false
        # All other exceptions will be caught.
      else
        true
      end
    end
  end

  # An Exception Tag (cf. Exceptional Ruby) that instructs Pry to show the error in
  # a more user-friendly manner. This should be used when the exception happens within
  # Pry itself as a direct consequence of the user typing something wrong.
  #
  # This allows us to distinguish between the user typing:
  #
  # pry(main)> def )
  # SyntaxError: unexpected )
  #
  # pry(main)> method_that_evals("def )")
  # SyntaxError: (eval):1: syntax error, unexpected ')'
  # from ./a.rb:2 in `eval'
  module UserError; end

  # Catches SecurityErrors if $SAFE is set
  module TooSafeException
    def self.===(exception)
      $SAFE > 0 && SecurityError === exception
    end
  end

  # When we try to get a binding for an object, we try to define a method on
  # that Object's singleton class. This doesn't work for "frozen" Object's, and
  # the exception is just a vanilla RuntimeError.
  module FrozenObjectException
    def self.===(exception)
      ["can't modify frozen class/module", "can't modify frozen Class"].include? exception.message
    end
  end

  # Don't catch these exceptions
  DEFAULT_EXCEPTION_WHITELIST = [SystemExit, SignalException, Pry::TooSafeException]

  # CommandErrors are caught by the REPL loop and displayed to the user. They
  # indicate an exceptional condition that's fatal to the current command.
  class CommandError < StandardError; end
  class MethodNotFound < CommandError; end

  # indicates obsolete API
  class ObsoleteError < StandardError; end

  # This is to keep from breaking under Rails 3.2 for people who are doing that
  # IRB = Pry thing.
  module ExtendCommandBundle
  end
end

if Pry::Helpers::BaseHelpers.mri_18?
  begin
  require 'ruby18_source_location'
  rescue LoadError
  end
end

require 'method_source'
require 'shellwords'
require 'stringio'
require 'coderay'
require 'slop'
require 'rbconfig'
require 'tempfile'
require 'pathname'

begin
  begin
    require 'readline'
  rescue LoadError
    require 'rb-readline'
  end
rescue LoadError
  warn "You're running a version of ruby with no Readline support"
  warn "Please `gem install rb-readline` or recompile ruby --with-readline."
  exit!
end

if Pry::Helpers::BaseHelpers.jruby?
  begin
    require 'ffi'
  rescue LoadError
    warn "Need to `gem install ffi`"
  end
end

if Pry::Helpers::BaseHelpers.windows? && !Pry::Helpers::BaseHelpers.windows_ansi?
  begin
    require 'win32console'
  # The mswin and mingw versions of pry require win32console, so this should
  # only fail on jruby (where win32console doesn't work).
  # Instead we'll recommend ansicon, which does.
  rescue LoadError
    warn "For a better pry experience, please use ansicon: http://adoxa.3eeweb.com/ansicon/"
  end
end

begin
  require 'bond'
rescue LoadError
end

require 'pry/version'
require 'pry/repl'
require 'pry/rbx_path'
require 'pry/code'
require 'pry/history_array'
require 'pry/helpers'
require 'pry/code_object'
require 'pry/method'
require 'pry/wrapped_module'
require 'pry/history'
require 'pry/command'
require 'pry/command_set'
require 'pry/commands'
require 'pry/custom_completions'
require 'pry/completion'
require 'pry/plugins'
require 'pry/core_extensions'
require 'pry/pry_class'
require 'pry/pry_instance'
require 'pry/cli'
require 'pry/pager'
require 'pry/terminal'
require 'pry/editor'
require 'pry/rubygem'
